File: packages/python/elizaos/features/advanced_capabilities/services/relationships.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from enum import StrEnum
from typing import TYPE_CHECKING
from uuid import UUID

from elizaos.types import Service, ServiceType

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime


class ContactCategoryEnum(StrEnum):
    FRIEND = "friend"
    FAMILY = "family"
    COLLEAGUE = "colleague"
    ACQUAINTANCE = "acquaintance"
    VIP = "vip"
    BUSINESS = "business"


@dataclass
class ContactCategory:
    id: str
    name: str
    description: str = ""
    color: str = ""


@dataclass
class ContactPreferences:
    preferred_channel: str | None = None
    timezone: str | None = None
    language: str | None = None
    contact_frequency: str | None = None
    do_not_disturb: bool = False
    notes: str | None = None


@dataclass
class ContactInfo:
    entity_id: UUID
    categories: list[str] = field(default_factory=list)
    tags: list[str] = field(default_factory=list)
    preferences: ContactPreferences = field(default_factory=ContactPreferences)
    custom_fields: dict[str, str | int | float | bool] = field(default_factory=dict)
    privacy_level: str = "private"
    last_modified: str = ""


@dataclass
class RelationshipAnalytics:
    strength: float = 0.0
    interaction_count: int = 0
    last_interaction_at: str | None = None
    average_response_time: float | None = None
    sentiment_score: float | None = None
    topics_discussed: list[str] = field(default_factory=list)


@dataclass
class RelationshipInsight:
    entity_id: str
    analytics: RelationshipAnalytics


@dataclass
class NeedsAttention:
    entity_id: str
    days_since_contact: int


@dataclass
class RelationshipInsights:
    strongest_relationships: list[RelationshipInsight] = field(default_factory=list)
    needs_attention: list[NeedsAttention] = field(default_factory=list)
    recent_interactions: list[RelationshipInsight] = field(default_factory=list)


def calculate_relationship_strength(
    interaction_count: int,
    last_interaction_at: str | None = None,
    message_quality: float = 5.0,
    relationship_type: str = "acquaintance",
) -> float:
    interaction_score = min(interaction_count * 2, 40)

    recency_score = 0.0
    if last_interaction_at:
        last_dt = datetime.fromisoformat(last_interaction_at.replace("Z", "+00:00"))
        days_since = (datetime.now(last_dt.tzinfo) - last_dt).days
        if days_since < 1:
            recency_score = 30
        elif days_since < 7:
            recency_score = 25
        elif days_since < 30:
            recency_score = 15
        elif days_since < 90:
            recency_score = 5

    quality_score = min(message_quality * 2, 20)

    relationship_bonus = {
        "family": 10,
        "friend": 8,
        "colleague": 6,
        "acquaintance": 4,
        "unknown": 0,
    }

    total = (
        interaction_score
        + recency_score
        + quality_score
        + relationship_bonus.get(relationship_type, 0)
    )
    return max(0.0, min(100.0, round(total, 1)))


class RelationshipsService(Service):
    name = "relationships"
    service_type = ServiceType.RELATIONSHIPS

    @property
    def capability_description(self) -> str:
        return "Comprehensive contact and relationship management service"

    _DEFAULT_CATEGORIES: list[ContactCategory] = [
        ContactCategory(id="friend", name="Friend", color="#4CAF50"),
        ContactCategory(id="family", name="Family", color="#2196F3"),
        ContactCategory(id="colleague", name="Colleague", color="#FF9800"),
        ContactCategory(id="acquaintance", name="Acquaintance", color="#9E9E9E"),
        ContactCategory(id="vip", name="VIP", color="#9C27B0"),
        ContactCategory(id="business", name="Business", color="#795548"),
    ]

    def __init__(self) -> None:
        self._contacts: dict[UUID, ContactInfo] = {}
        self._analytics: dict[str, RelationshipAnalytics] = {}
        self._categories: list[ContactCategory] = list(self._DEFAULT_CATEGORIES)
        self._runtime: IAgentRuntime | None = None

    @classmethod
    async def start(cls, runtime: IAgentRuntime) -> RelationshipsService:
        service = cls()
        service._runtime = runtime
        runtime.logger.info(
            "Relationships service started",
            src="service:relationships",
            agentId=str(runtime.agent_id),
        )
        return service

    async def stop(self) -> None:
        if self._runtime:
            self._runtime.logger.info(
                "Relationships service stopped",
                src="service:relationships",
                agentId=str(self._runtime.agent_id),
            )
        self._contacts.clear()
        self._analytics.clear()
        self._categories.clear()
        self._runtime = None

    async def add_contact(
        self,
        entity_id: UUID,
        categories: list[str] | None = None,
        preferences: ContactPreferences | None = None,
        custom_fields: dict[str, str | int | float | bool] | None = None,
    ) -> ContactInfo:
        contact = ContactInfo(
            entity_id=entity_id,
            categories=categories or ["acquaintance"],
            tags=[],
            preferences=preferences or ContactPreferences(),
            custom_fields=custom_fields or {},
            privacy_level="private",
            last_modified=datetime.utcnow().isoformat(),
        )

        self._contacts[entity_id] = contact

        if self._runtime:
            self._runtime.logger.info(
                f"Added contact {entity_id}",
                src="service:relationships",
                categories=contact.categories,
            )

        return contact

    async def get_contact(self, entity_id: UUID) -> ContactInfo | None:
        return self._contacts.get(entity_id)

    async def update_contact(
        self,
        entity_id: UUID,
        categories: list[str] | None = None,
        tags: list[str] | None = None,
        preferences: ContactPreferences | None = None,
        custom_fields: dict[str, str | int | float | bool] | None = None,
    ) -> ContactInfo | None:
        contact = self._contacts.get(entity_id)
        if not contact:
            return None

        if categories is not None:
            contact.categories = categories
        if tags is not None:
            contact.tags = tags
        if preferences is not None:
            contact.preferences = preferences
        if custom_fields is not None:
            contact.custom_fields = custom_fields

        contact.last_modified = datetime.utcnow().isoformat()

        return contact

    async def remove_contact(self, entity_id: UUID) -> bool:
        if entity_id in self._contacts:
            del self._contacts[entity_id]
            return True
        return False

    async def search_contacts(
        self,
        categories: list[str] | None = None,
        tags: list[str] | None = None,
        search_term: str | None = None,
    ) -> list[ContactInfo]:
        results = list(self._contacts.values())

        if categories:
            results = [c for c in results if any(cat in c.categories for cat in categories)]

        if tags:
            results = [c for c in results if any(tag in c.tags for tag in tags)]

        normalized_search = (search_term or "").strip().lower()
        if normalized_search:
            filtered: list[ContactInfo] = []
            for contact in results:
                if normalized_search in str(contact.entity_id).lower():
                    filtered.append(contact)
                    continue

                entity = None
                if self._runtime is not None:
                    entity = await self._runtime.get_entity(str(contact.entity_id))

                entity_name = getattr(entity, "name", None)
                if isinstance(entity_name, str) and normalized_search in entity_name.lower():
                    filtered.append(contact)
                    continue

                entity_names = getattr(entity, "names", None)
                if isinstance(entity_names, list) and any(
                    isinstance(name, str) and normalized_search in name.lower()
                    for name in entity_names
                ):
                    filtered.append(contact)

            results = filtered

        return results

    async def get_all_contacts(self) -> list[ContactInfo]:
        return list(self._contacts.values())

    async def get_relationship_analytics(
        self,
        entity_id: UUID,
    ) -> RelationshipAnalytics | None:
        key = str(entity_id)
        return self._analytics.get(key)

    async def update_relationship_analytics(
        self,
        entity_id: UUID,
        interaction_count: int | None = None,
        last_interaction_at: str | None = None,
    ) -> RelationshipAnalytics:
        key = str(entity_id)
        analytics = self._analytics.get(key) or RelationshipAnalytics()

        if interaction_count is not None:
            analytics.interaction_count = interaction_count
        if last_interaction_at is not None:
            analytics.last_interaction_at = last_interaction_at

        contact = self._contacts.get(entity_id)
        relationship_type = "acquaintance"
        if contact and contact.categories:
            relationship_type = contact.categories[0]

        analytics.strength = calculate_relationship_strength(
            analytics.interaction_count,
            analytics.last_interaction_at,
            relationship_type=relationship_type,
        )

        self._analytics[key] = analytics
        return analytics

    # ------------------------------------------------------------------
    # Relationship analysis
    # ------------------------------------------------------------------

    async def analyze_relationship(
        self,
        source_entity_id: str,
        target_entity_id: str,
    ) -> RelationshipAnalytics | None:
        """Analyze the relationship between two specific entities.

        Looks up the analytics stored for *source* -> *target* (or
        *target* -> *source*).  Returns ``None`` when no analytics exist
        for either direction.
        """
        for key in (
            f"{source_entity_id}-{target_entity_id}",
            f"{target_entity_id}-{source_entity_id}",
        ):
            analytics = self._analytics.get(key)
            if analytics is not None:
                return analytics
        return None

    async def get_relationship_insights(
        self,
        entity_id: str,
    ) -> RelationshipInsights:
        """Return categorized insights for all relationships of *entity_id*.

        * **strongest_relationships** -- top 10 by strength (descending).
        * **needs_attention** -- contacts with no interaction in 30+ days,
          sorted by days since contact descending.
        * **recent_interactions** -- last 10 by timestamp (most recent first).
        """
        strongest: list[RelationshipInsight] = []
        attention: list[NeedsAttention] = []
        recent: list[tuple[str, RelationshipInsight]] = []  # (iso-ts, insight)

        now = datetime.utcnow()

        for key, analytics in self._analytics.items():
            parts = key.split("-", 1)
            if len(parts) != 2:
                continue

            # Only consider analytics that involve *entity_id*.
            if entity_id not in parts:
                continue

            other_id = parts[1] if parts[0] == entity_id else parts[0]
            insight = RelationshipInsight(entity_id=other_id, analytics=analytics)

            strongest.append(insight)

            if analytics.last_interaction_at:
                last_dt = datetime.fromisoformat(
                    analytics.last_interaction_at.replace("Z", "+00:00")
                )
                # Make *now* offset-aware when the stored timestamp is.
                ref = now if last_dt.tzinfo is None else now.replace(tzinfo=last_dt.tzinfo)
                days_since = (ref - last_dt).days
                if days_since >= 30:
                    attention.append(
                        NeedsAttention(entity_id=other_id, days_since_contact=days_since)
                    )
                recent.append((analytics.last_interaction_at, insight))

        # Top 10 strongest by strength descending.
        strongest.sort(key=lambda i: i.analytics.strength, reverse=True)
        strongest = strongest[:10]

        # Needs attention sorted by staleness descending.
        attention.sort(key=lambda a: a.days_since_contact, reverse=True)

        # Last 10 recent interactions by timestamp descending.
        recent.sort(key=lambda pair: pair[0], reverse=True)
        recent_insights = [pair[1] for pair in recent[:10]]

        return RelationshipInsights(
            strongest_relationships=strongest,
            needs_attention=attention,
            recent_interactions=recent_insights,
        )

    # ------------------------------------------------------------------
    # Category management
    # ------------------------------------------------------------------

    async def get_categories(self) -> list[ContactCategory]:
        """Return all contact categories."""
        return list(self._categories)

    async def add_category(self, category: ContactCategory) -> None:
        """Add a new category.

        Raises ``ValueError`` if a category with the same *id* already
        exists.
        """
        if any(c.id == category.id for c in self._categories):
            raise ValueError(f"Category '{category.id}' already exists")
        self._categories.append(category)
        if self._runtime:
            self._runtime.logger.info(
                f"Added category {category.name}",
                src="service:relationships",
            )

    # ------------------------------------------------------------------
    # Privacy management
    # ------------------------------------------------------------------

    async def set_contact_privacy(
        self,
        entity_id: str,
        privacy_level: str,
    ) -> bool:
        """Set the privacy level on a contact.

        Returns ``True`` on success, ``False`` if the contact does not
        exist.
        """
        if privacy_level not in ("public", "private", "restricted"):
            return False
        uid = UUID(entity_id) if isinstance(entity_id, str) else entity_id
        contact = self._contacts.get(uid)
        if contact is None:
            return False
        contact.privacy_level = privacy_level
        contact.last_modified = datetime.utcnow().isoformat()
        if self._runtime:
            self._runtime.logger.info(
                f"Set privacy for {entity_id} to {privacy_level}",
                src="service:relationships",
            )
        return True

    async def can_access_contact(
        self,
        requesting_entity_id: str,
        target_entity_id: str,
    ) -> bool:
        """Check whether *requesting_entity_id* may access *target_entity_id*.

        Access rules based on the target contact's ``privacy_level``:

        * ``"public"``     -- always accessible.
        * ``"private"``    -- accessible only to the entity itself.
        * ``"restricted"`` -- agent-only (returns ``False`` for all
          external callers; the owning agent bypasses this check).
        """
        uid = UUID(target_entity_id) if isinstance(target_entity_id, str) else target_entity_id
        contact = self._contacts.get(uid)
        if contact is None:
            return False

        level = contact.privacy_level

        if level == "public":
            return True

        if level == "private":
            return requesting_entity_id == target_entity_id

        # "restricted" — agent-only; non-agent callers always denied.
        return False


RelationshipsService = RelationshipsService

File: packages/rust/src/features/basic_capabilities/bootstrap/services/relationships.rs
//! Relationships service implementation.

use async_trait::async_trait;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Arc;
use uuid::Uuid;

use crate::error::PluginResult;
use crate::runtime::IAgentRuntime;

use super::{Service, ServiceType};

/// A named contact category (e.g., "friend", "family", "vip").
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContactCategory {
    pub id: String,
    pub name: String,
    pub description: String,
    pub color: String,
}

/// Categorized relationship insights for an entity.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct RelationshipInsights {
    /// Top relationships sorted by strength (up to 10).
    pub strongest_relationships: Vec<RelationshipInsightEntry>,
    /// Contacts with no interaction in 30+ days.
    pub needs_attention: Vec<NeedsAttentionEntry>,
    /// Most recent interactions (up to 10, newest first).
    pub recent_interactions: Vec<RecentInteractionEntry>,
}

/// A single entry in the "strongest relationships" list.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RelationshipInsightEntry {
    pub entity_id: Uuid,
    pub analytics: RelationshipAnalytics,
}

/// A contact that has not been interacted with recently.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NeedsAttentionEntry {
    pub entity_id: Uuid,
    pub days_since_contact: i64,
}

/// A recently-interacted contact.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RecentInteractionEntry {
    pub entity_id: Uuid,
    pub last_interaction: DateTime<Utc>,
}

/// Contact preferences.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ContactPreferences {
    pub preferred_channel: Option<String>,
    pub timezone: Option<String>,
    pub language: Option<String>,
    pub contact_frequency: Option<String>,
    pub do_not_disturb: bool,
    pub notes: Option<String>,
}

/// Contact information.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContactInfo {
    pub entity_id: Uuid,
    pub categories: Vec<String>,
    pub tags: Vec<String>,
    pub preferences: ContactPreferences,
    pub custom_fields: HashMap<String, serde_json::Value>,
    pub privacy_level: String,
    pub last_modified: DateTime<Utc>,
}

/// Relationship analytics.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct RelationshipAnalytics {
    pub strength: f64,
    pub interaction_count: u32,
    pub last_interaction_at: Option<DateTime<Utc>>,
    pub average_response_time: Option<f64>,
    pub sentiment_score: Option<f64>,
    pub topics_discussed: Vec<String>,
}

/// Calculate relationship strength based on interaction patterns.
pub fn calculate_relationship_strength(
    interaction_count: u32,
    last_interaction_at: Option<DateTime<Utc>>,
    message_quality: f64,
    relationship_type: &str,
) -> f64 {
    // Base score from interaction count (max 40 points)
    let interaction_score = (interaction_count as f64 * 2.0).min(40.0);

    // Recency score (max 30 points)
    let recency_score = if let Some(last) = last_interaction_at {
        let days_since = (Utc::now() - last).num_days();
        if days_since < 1 {
            30.0
        } else if days_since < 7 {
            25.0
        } else if days_since < 30 {
            15.0
        } else if days_since < 90 {
            5.0
        } else {
            0.0
        }
    } else {
        0.0
    };

    // Quality score (max 20 points)
    let quality_score = (message_quality * 2.0).min(20.0);

    // Relationship type bonus (max 10 points)
    let relationship_bonus = match relationship_type {
        "family" => 10.0,
        "friend" => 8.0,
        "colleague" => 6.0,
        "acquaintance" => 4.0,
        _ => 0.0,
    };

    let total = interaction_score + recency_score + quality_score + relationship_bonus;
    total.max(0.0).min(100.0)
}

/// Service for managing contacts and relationships.
pub struct RelationshipsService {
    contacts: HashMap<Uuid, ContactInfo>,
    analytics: HashMap<Uuid, RelationshipAnalytics>,
    categories: Vec<ContactCategory>,
    runtime: Option<Arc<dyn IAgentRuntime>>,
}

impl RelationshipsService {
    /// Create a new relationships service.
    pub fn new() -> Self {
        Self {
            contacts: HashMap::new(),
            analytics: HashMap::new(),
            categories: default_categories(),
            runtime: None,
        }
    }

    /// Add a new contact.
    pub fn add_contact(
        &mut self,
        entity_id: Uuid,
        categories: Vec<String>,
        preferences: Option<ContactPreferences>,
    ) -> ContactInfo {
        let contact = ContactInfo {
            entity_id,
            categories: if categories.is_empty() {
                vec!["acquaintance".to_string()]
            } else {
                categories
            },
            tags: Vec::new(),
            preferences: preferences.unwrap_or_default(),
            custom_fields: HashMap::new(),
            privacy_level: "private".to_string(),
            last_modified: Utc::now(),
        };

        if let Some(runtime) = &self.runtime {
            runtime.log_info(
                "service:relationships",
                &format!("Added contact: {}", entity_id),
            );
        }

        self.contacts.insert(entity_id, contact.clone());
        contact
    }

    /// Get a contact by entity ID.
    pub fn get_contact(&self, entity_id: Uuid) -> Option<&ContactInfo> {
        self.contacts.get(&entity_id)
    }

    /// Update a contact.
    pub fn update_contact(
        &mut self,
        entity_id: Uuid,
        categories: Option<Vec<String>>,
        tags: Option<Vec<String>>,
    ) -> Option<&ContactInfo> {
        if let Some(contact) = self.contacts.get_mut(&entity_id) {
            if let Some(cats) = categories {
                contact.categories = cats;
            }
            if let Some(t) = tags {
                contact.tags = t;
            }
            contact.last_modified = Utc::now();
            Some(contact)
        } else {
            None
        }
    }

    /// Remove a contact.
    pub fn remove_contact(&mut self, entity_id: Uuid) -> bool {
        self.contacts.remove(&entity_id).is_some()
    }

    /// Search contacts by criteria.
    pub fn search_contacts(
        &self,
        categories: Option<&[String]>,
        tags: Option<&[String]>,
    ) -> Vec<&ContactInfo> {
        self.contacts
            .values()
            .filter(|c| {
                let cat_match = categories
                    .map(|cats| cats.iter().any(|cat| c.categories.contains(cat)))
                    .unwrap_or(true);
                let tag_match = tags
                    .map(|ts| ts.iter().any(|t| c.tags.contains(t)))
                    .unwrap_or(true);
                cat_match && tag_match
            })
            .collect()
    }

    /// Get all contacts.
    pub fn get_all_contacts(&self) -> Vec<&ContactInfo> {
        self.contacts.values().collect()
    }

    /// Get relationship analytics.
    pub fn get_analytics(&self, entity_id: Uuid) -> Option<&RelationshipAnalytics> {
        self.analytics.get(&entity_id)
    }

    /// Update relationship analytics.
    pub fn update_analytics(
        &mut self,
        entity_id: Uuid,
        interaction_count: Option<u32>,
        last_interaction_at: Option<DateTime<Utc>>,
    ) -> &RelationshipAnalytics {
        let analytics = self.analytics.entry(entity_id).or_default();

        if let Some(count) = interaction_count {
            analytics.interaction_count = count;
        }
        if let Some(last) = last_interaction_at {
            analytics.last_interaction_at = Some(last);
        }

        // Recalculate strength
        let relationship_type = self
            .contacts
            .get(&entity_id)
            .and_then(|c| c.categories.first())
            .map(|s| s.as_str())
            .unwrap_or("acquaintance");

        analytics.strength = calculate_relationship_strength(
            analytics.interaction_count,
            analytics.last_interaction_at,
            5.0,
            relationship_type,
        );

        analytics
    }

    /// Analyze the relationship between two entities, returning computed
    /// analytics (strength, interaction count, sentiment, topics).
    ///
    /// Returns `None` if neither entity has analytics recorded.
    pub fn analyze_relationship(
        &self,
        source_entity_id: &Uuid,
        target_entity_id: &Uuid,
    ) -> Option<RelationshipAnalytics> {
        // Check the composite key first, then fall back to per-entity analytics.
        let composite_key = composite_analytics_key(source_entity_id, target_entity_id);
        if let Some(a) = self.analytics.get(&composite_key) {
            return Some(a.clone());
        }

        // Fall back: merge data from both sides.
        let source = self.analytics.get(source_entity_id);
        let target = self.analytics.get(target_entity_id);

        match (source, target) {
            (Some(s), Some(t)) => {
                let interaction_count = s.interaction_count + t.interaction_count;
                let last_interaction_at = match (s.last_interaction_at, t.last_interaction_at) {
                    (Some(a), Some(b)) => Some(a.max(b)),
                    (a @ Some(_), None) | (None, a @ Some(_)) => a,
                    (None, None) => None,
                };
                let sentiment_score = match (s.sentiment_score, t.sentiment_score) {
                    (Some(a), Some(b)) => Some((a + b) / 2.0),
                    (a, b) => a.or(b),
                };
                let mut topics: Vec<String> = s.topics_discussed.clone();
                for topic in &t.topics_discussed {
                    if !topics.contains(topic) {
                        topics.push(topic.clone());
                    }
                }
                let relationship_type = self
                    .contacts
                    .get(source_entity_id)
                    .or_else(|| self.contacts.get(target_entity_id))
                    .and_then(|c| c.categories.first())
                    .map(|s| s.as_str())
                    .unwrap_or("acquaintance");
                let strength = calculate_relationship_strength(
                    interaction_count,
                    last_interaction_at,
                    5.0,
                    relationship_type,
                );
                Some(RelationshipAnalytics {
                    strength,
                    interaction_count,
                    last_interaction_at,
                    average_response_time: s.average_response_time.or(t.average_response_time),
                    sentiment_score,
                    topics_discussed: topics,
                })
            }
            (Some(a), None) | (None, Some(a)) => Some(a.clone()),
            (None, None) => None,
        }
    }

    /// Return categorized relationship insights for an entity.
    ///
    /// * **strongest_relationships** -- top 10 contacts by strength score.
    /// * **needs_attention** -- contacts with no interaction in 30+ days.
    /// * **recent_interactions** -- last 10 contacts by interaction time.
    pub fn get_relationship_insights(&self, entity_id: &Uuid) -> RelationshipInsights {
        let now = Utc::now();

        // Collect analytics for every *other* entity that shares analytics with `entity_id`.
        let mut entries: Vec<(Uuid, RelationshipAnalytics)> = Vec::new();
        for (other_id, analytics) in &self.analytics {
            if other_id == entity_id {
                continue;
            }
            entries.push((*other_id, analytics.clone()));
        }

        // Strongest (top 10 by strength, descending).
        let mut by_strength = entries.clone();
        by_strength.sort_by(|a, b| {
            b.1.strength
                .partial_cmp(&a.1.strength)
                .unwrap_or(std::cmp::Ordering::Equal)
        });
        let strongest_relationships: Vec<RelationshipInsightEntry> = by_strength
            .iter()
            .take(10)
            .map(|(id, a)| RelationshipInsightEntry {
                entity_id: id.clone(),
                analytics: a.clone(),
            })
            .collect();

        // Needs attention (no interaction in 30+ days).
        let needs_attention: Vec<NeedsAttentionEntry> = entries
            .iter()
            .filter_map(|(id, a)| {
                a.last_interaction_at.and_then(|last| {
                    let days = (now - last).num_days();
                    if days >= 30 {
                        Some(NeedsAttentionEntry {
                            entity_id: id.clone(),
                            days_since_contact: days,
                        })
                    } else {
                        None
                    }
                })
            })
            .collect();

        // Recent interactions (last 10, newest first).
        let mut with_interaction: Vec<(Uuid, DateTime<Utc>)> = entries
            .iter()
            .filter_map(|(id, a)| a.last_interaction_at.map(|ts| (id.clone(), ts)))
            .collect();
        with_interaction.sort_by(|a, b| b.1.cmp(&a.1));
        let recent_interactions: Vec<RecentInteractionEntry> = with_interaction
            .into_iter()
            .take(10)
            .map(|(id, ts)| RecentInteractionEntry {
                entity_id: id,
                last_interaction: ts,
            })
            .collect();

        RelationshipInsights {
            strongest_relationships,
            needs_attention,
            recent_interactions,
        }
    }

    /// Return all contact categories.
    pub fn get_categories(&self) -> &[ContactCategory] {
        &self.categories
    }

    /// Add a new contact category. Does nothing if a category with the same
    /// `id` already exists.
    pub fn add_category(&mut self, category: ContactCategory) {
        if self.categories.iter().any(|c| c.id == category.id) {
            return;
        }
        if let Some(runtime) = &self.runtime {
            runtime.log_info(
                "service:relationships",
                &format!("Added category: {}", category.name),
            );
        }
        self.categories.push(category);
    }

    /// Set the privacy level on a contact. Returns `false` if the contact does
    /// not exist.
    pub fn set_contact_privacy(&mut self, entity_id: &Uuid, privacy_level: &str) -> bool {
        let valid = matches!(privacy_level, "public" | "private" | "restricted");
        if !valid {
            return false;
        }
        if let Some(contact) = self.contacts.get_mut(entity_id) {
            contact.privacy_level = privacy_level.to_string();
            contact.last_modified = Utc::now();
            if let Some(runtime) = &self.runtime {
                runtime.log_info(
                    "service:relationships",
                    &format!("Set privacy for {} to {}", entity_id, privacy_level),
                );
            }
            true
        } else {
            false
        }
    }

    /// Check whether `requesting_entity_id` can access the contact record of
    /// `target_entity_id`. The owning agent always has access.
    pub fn can_access_contact(&self, requesting_entity_id: &Uuid, target_entity_id: &Uuid) -> bool {
        let contact = match self.contacts.get(target_entity_id) {
            Some(c) => c,
            None => return false,
        };
        match contact.privacy_level.as_str() {
            "public" => true,
            "private" => requesting_entity_id == target_entity_id,
            "restricted" => false,
            _ => false,
        }
    }
}

/// Build a deterministic composite key for a pair of entity IDs so that
/// `(a, b)` and `(b, a)` map to the same analytics entry.
fn composite_analytics_key(a: &Uuid, b: &Uuid) -> Uuid {
    use std::hash::{Hash, Hasher};
    let (lo, hi) = if a <= b { (a, b) } else { (b, a) };
    let mut hasher = std::collections::hash_map::DefaultHasher::new();
    lo.hash(&mut hasher);
    hi.hash(&mut hasher);
    let hash = hasher.finish();
    // Construct a v4-shaped UUID from the hash (not cryptographic, just a key).
    let bytes = hash.to_le_bytes();
    let mut buf = [0u8; 16];
    buf[..8].copy_from_slice(&bytes);
    buf[8..16].copy_from_slice(&bytes);
    Uuid::from_bytes(buf)
}

/// Default contact categories matching the TypeScript service.
fn default_categories() -> Vec<ContactCategory> {
    vec![
        ContactCategory {
            id: "friend".into(),
            name: "Friend".into(),
            description: String::new(),
            color: "#4CAF50".into(),
        },
        ContactCategory {
            id: "family".into(),
            name: "Family".into(),
            description: String::new(),
            color: "#2196F3".into(),
        },
        ContactCategory {
            id: "colleague".into(),
            name: "Colleague".into(),
            description: String::new(),
            color: "#FF9800".into(),
        },
        ContactCategory {
            id: "acquaintance".into(),
            name: "Acquaintance".into(),
            description: String::new(),
            color: "#9E9E9E".into(),
        },
        ContactCategory {
            id: "vip".into(),
            name: "VIP".into(),
            description: String::new(),
            color: "#9C27B0".into(),
        },
        ContactCategory {
            id: "business".into(),
            name: "Business".into(),
            description: String::new(),
            color: "#795548".into(),
        },
    ]
}

impl Default for RelationshipsService {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl Service for RelationshipsService {
    fn name(&self) -> &'static str {
        "relationships"
    }

    fn service_type(&self) -> ServiceType {
        ServiceType::Core
    }

    async fn start(&mut self, runtime: Arc<dyn IAgentRuntime>) -> PluginResult<()> {
        runtime.log_info("service:relationships", "Relationships service started");
        self.runtime = Some(runtime);
        Ok(())
    }

    async fn stop(&mut self) -> PluginResult<()> {
        if let Some(runtime) = &self.runtime {
            runtime.log_info("service:relationships", "Relationships service stopped");
        }
        self.contacts.clear();
        self.analytics.clear();
        self.categories.clear();
        self.runtime = None;
        Ok(())
    }
}
